aws/rds_snapshot.go

```diff
@@ -0,0 +1,191 @@
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/rds"
+	"github.com/gruntwork-io/cloud-nuke/config"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+// Get all DB Instance snapshots
+func getAllRdsSnapshots(session *session.Session, excludeAfter time.Time, configObj config.Config) ([]*string, error) {
+	svc := rds.New(session)
+
+	result, err := svc.DescribeDBSnapshots(&rds.DescribeDBSnapshotsInput{})
+
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	var snapshots []*string
+
+	for _, snapshot := range result.DBSnapshots {
+
+		// List all DB Instance Snapshot tags
+		tagsResult, err := svc.ListTagsForResource(&rds.ListTagsForResourceInput{
+			ResourceName: snapshot.DBSnapshotArn,
+		})
+
+		if err != nil {
+			return nil, errors.WithStackTrace(err)
+		}
+
+		// Automated DB snapshots can only be deleted by deleting the DB instance or
+		// changing the backup retention period for the DB instance to 0.
+		// This edge case can't be handled since all DB instance related automated snapshots will be deleted.
+		// Refer to https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_DeleteSnapshot.html
+		if *snapshot.SnapshotType != "automated" {
+			if snapshot.SnapshotCreateTime != nil && excludeAfter.After(awsgo.TimeValue(snapshot.SnapshotCreateTime)) {
+				if shouldIncludeSnapshotByName(*snapshot.DBSnapshotIdentifier, configObj.RDSSnapshots.IncludeRule.NamesRE, configObj.RDSSnapshots.ExcludeRule.NamesRE) {
+					if len(tagsResult.TagList) > 0 {
+						for _, tag := range tagsResult.TagList {
+							if shouldIncludeSnapshotByTag(*tag.Key, configObj.RDSSnapshots.IncludeRule.TagNamesRE, configObj.RDSSnapshots.ExcludeRule.TagNamesRE) {
+								snapshots = append(snapshots, snapshot.DBSnapshotIdentifier)
+							}
+						}
+					} else {
+						snapshots = append(snapshots, snapshot.DBSnapshotIdentifier)
+					}
+				}
+			}
+		}
+	}
+
+	return snapshots, nil
+}
+
+// Filter DB Instance snapshot by names_regex in config file
+func shouldIncludeSnapshotByName(snapshotName string, includeNamesREList []*regexp.Regexp, excludeNamesREList []*regexp.Regexp) bool {
+	shouldInclude := false
+
+	if len(includeNamesREList) > 0 {
+		// If any include rules are defined
+		// and the include rule matches the snapshot, check to see if an exclude rule matches
+		if includeSnapshotByNamesREList(snapshotName, includeNamesREList) {
+			shouldInclude = excludeSnapshotByNamesREList(snapshotName, excludeNamesREList)
+		}
+	} else if len(excludeNamesREList) > 0 {
+		// If there are no include rules defined, check to see if an exclude rule matches
+		shouldInclude = excludeSnapshotByNamesREList(snapshotName, excludeNamesREList)
+	} else {
+		// Otherwise
+		shouldInclude = true
+	}
+
+	return shouldInclude
+}
+
+// Include filtered DB Instance snapshot
+func includeSnapshotByNamesREList(snapshotName string, reList []*regexp.Regexp) bool {
```

Comment:
Same comment on a single function for handling these checks

Reply:
Fixed.